magento-research/pwa-studio
# enhancement: unify fastly and local media URLs (no more 301s for images in production)

## Description

Simplify the creation and consumption of image URLs, so they work the
same way in all environments.

This removes the extra 301 redirects to optimized images in production.

- Add a "preload" data to the HTML tag, which renders the Magento
backend URL and the image optimization source (`backend` or `origin`) as data attributes
- Modify `makeUrl()` to query those attributes and:
    - if `IMAGE_OPTIMIZING_ORIGIN` is `backend`, indicating Fastly or any other optimization service is enabled on the backend, then force ALL media URLs to use the backend as the base URL and render as absolute.
   - if `IMAGE_OPTIMIZING_ORIGIN` is `onboard`, indicating the Web server must optimize images, then force ALL media URLs to to render as relative.
- Modify the backend `addOptImgMiddleware()` feature to work with Fastly-style URL parameters, replicating a subset of its features.
- Remove image optimization logic from UPWARD entirely, since it is now
delegated to other layers.

## Related Issue





Resolves the ongoing performance implication of 301 redirects in production.
Resolves the HTTP/2 redirect ding that Lighthouse gives the changes in #1165 .

## Verification Steps

1. Run a staging or development server.
1. Open the network tab in dev tools. Filter for images.
1. Go to the homepage.
1. Observe that the images are compressed and that they have headers indicating that PWA Studio runtime compressed them.
1. Visit a production cloud server running Fastly IO.
1. Go to the homepage.
1. Observe that the image URLs have the same structure, but their headers indicate they were compressed with Fastly instead.

## How Have YOU Tested this?
In Unit Tests: Updated unit tests for makeURL, and updated the snapshots which included makeURL output
For UAT: Locally, but not in Fastly IO yet. Since we let those URLs through unmolested, it should work great.

## Screenshots / Screen Captures (if appropriate):

## Proposed Labels for Change Type/Package




## Checklist:
